From bd896e6ccd8d9bb4a8adb1b89192c7c61e9d04ac Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Wed, 18 Jun 2014 10:27:58 -0400
Subject: [s390] uaccess: introduce 'uaccesspt' kernel parameter

Message-id: <1403087280-902-13-git-send-email-brueckner@redhat.com>
Patchwork-id: 84249
O-Subject: [RHEL7.1 PATCH 12/14] [s390] s390/uaccess: introduce 'uaccesspt' kernel parameter
Bugzilla: 1099147
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Description
-----------
commit 56f15e518cfdc732bd4e4da90e0c9cf2fc4e7c1b
Author: Heiko Carstens <heiko.carstens@de.ibm.com>
Commit: Martin Schwidefsky <schwidefsky@de.ibm.com>

    s390/uaccess: introduce 'uaccesspt' kernel parameter

    The uaccesspt kernel parameter allows to enforce using the uaccess page
    table walk variant. This is mainly for debugging purposes, so this mode
    can also be enabled on machines which support the mvcos instruction.

    Signed-off-by: Heiko Carstens <heiko.carstens@de.ibm.com>
    Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

Bugzilla
--------
BZ 1099147
https://bugzilla.redhat.com/show_bug.cgi?id=1099147

Upstream status of the patch
----------------------------
The patch is upstream as of kernel version v3.15-rc1
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=56f15e518cfdc732bd4e4da90e0c9cf2fc4e7c1b

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=7588402

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/s390/lib/uaccess_mvcos.c b/arch/s390/lib/uaccess_mvcos.c
index e2685ff..ae97b8d 100644
--- a/arch/s390/lib/uaccess_mvcos.c
+++ b/arch/s390/lib/uaccess_mvcos.c
@@ -241,9 +241,22 @@ long __strncpy_from_user(char *dst, const char __user *src, long count)
 }
 EXPORT_SYMBOL(__strncpy_from_user);
 
+/*
+ * The uaccess page tabe walk variant can be enforced with the "uaccesspt"
+ * kernel parameter. This is mainly for debugging purposes.
+ */
+static int force_uaccess_pt __initdata;
+
+static int __init parse_uaccess_pt(char *__unused)
+{
+ force_uaccess_pt = 1;
+ return 0;
+}
+early_param("uaccesspt", parse_uaccess_pt);
+
 static int __init uaccess_init(void)
 {
- if (IS_ENABLED(CONFIG_32BIT) || !test_facility(27))
+ if (IS_ENABLED(CONFIG_32BIT) || force_uaccess_pt || !test_facility(27))
   static_key_slow_dec(&have_mvcos);
  return 0;
 }
-- 
1.7.1